src/nunavut/lang/c/support/serialization.j2

```diff
@@ -0,0 +1,556 @@
+#ifndef NUNAVUT_ASSERT
+// By default Nunavut does not generate assert statements since the logic to halt a program is platform
+// dependent and because this header requires an absolute minimum from a platform and from the C standard
+// library.
+// Most platforms can simply define NUNAVUT_ASSERT as assert from assert.h
+#   error "You must either define NUNAVUT_ASSERT or you need to disable assertions"
+          " in the serializer using language options"
+#endif
+{% endif -%}
+
+// --------------------------------------------- PRIMITIVE SERIALIZATION ---------------------------------------------
+
+/// Calculate the number of bits to safely copy from serialized buffer.
+/// Returns the smallest bit length based on requested parameters and available space.
+///
+/// @param buf_size_bytes       Size of the input buffer in bytes.
+/// @param offset_bit           Offset (in bits) of data type in message.
+/// @param requested_length_bit Size (in bits) of the data type requested.
+/// @param value_length_bit     Size (in bits) of the data type that is being extracted into.
+///
+/// @returns                    Number of bits to safely read from buffer.
+static inline {{typename_unsigned_bit_length}} nunavutInternalGetBitCopySize(
+    const size_t                   buf_size_bytes,
+    const {{typename_unsigned_bit_length}} offset_bit,
+    const {{typename_unsigned_bit_length}} requested_length_bit,
+    const uint8_t                  value_length_bit)
+{
+    const {{typename_unsigned_bit_length}} buf_size_bit  = ({{typename_unsigned_bit_length}})buf_size_bytes * 8U;
+    const {{typename_unsigned_bit_length}} remaining_bit = buf_size_bit - {{ choose_min('buf_size_bit', 'offset_bit') }};
+    const {{typename_unsigned_bit_length}} min_length_bit = {{ choose_min('requested_length_bit', 'value_length_bit') }};
+    const {{typename_unsigned_bit_length}} min_remaining_length = {{ choose_min('remaining_bit', 'min_length_bit') }};
+    return min_remaining_length;
+}
+
+// --------------------------------------------- PUBLIC API - BIT ARRAY ---------------------------------------------
+
+static inline void nunavutCopyBits(const {{typename_unsigned_bit_length}} length_bit,
+                                   const {{typename_unsigned_bit_length}} src_offset_bit,
+                                   const {{typename_unsigned_bit_length}} dst_offset_bit,
+                                   const uint8_t* const           src,
+                                   uint8_t* const                 dst)
+{
+    // The algorithm was originally designed by Ben Dyer for Libuavcan v0:
+    // https://github.com/UAVCAN/libuavcan/blob/ba6929f9625d7ea3eb00/libuavcan/src/marshal/uc_bit_array_copy.cpp#L12-L58
+    // This version is modified for v1 where the bit order is the opposite.
+{%- if option_enable_serialization_asserts %}
+    NUNAVUT_ASSERT((src != NULL) && (dst != NULL) && (src != dst));
+    NUNAVUT_ASSERT((src < dst)
```

Comment:
why +8u here? This means the assert would fail for:

```
    uint8_t dst[5];
    const uint8_t src[] = { 0, 1, 2, 3, 4 };

    nunavutCopyBits(5 * 8, 0, 0, src, dst);
```

but not

```
    uint8_t dst[5];
    volatile uint8_t foo = 5;
    const uint8_t src[] = { 0, 1, 2, 3, 4 };

    nunavutCopyBits(5 * 8, 0, 0, src, dst);
```

Reply:
I believe the intention was to round up, as this example would fail without any rounding (I think, haven't tested it):

```C
uint8_t buf[2];
nunavutCopyBits(8, 4, 4, &buf[0], &buf[1]);
```

For now I'm just going to change it to `+ 4U`, because I think that makes sense. I'm coming back at this later tonight.